### PR TITLE
Add commands `reset` and `hold-in-reset`

### PR DESCRIPTION
**Pull request by jnross**

### CODE REVIEW

File: espflash/src/bin/espflash.rs [hunk 1]
```diff
@@ -196,6 +202,23 @@ fn erase_region(args: EraseRegionArgs, config: &Config) -> Result<()> {
...
+    let mut args = args.clone();
+    args.no_stub = true;
+    let mut flash = connect(&args, config)?;
+    info!("Resetting target device");
+    flash.connection().reset()?;
+
+    Ok(())
+}
+
+fn hold_in_reset(args: ConnectArgs, config: &Config) -> Result<()> {
```
Comment (bobsjune) [inline espflash/src/bin/espflash.rs:215]: Thank you for naming this function helpfully!  I have to remind folks that the default state for an MCU is to be running, and that is has to be held in reset (or powered down) to NOT run.  So when we "reset" the MCU, we are actually "releasing from reset".  This function name makes that very explicit.
That said, I'm not requesting that you change the name of `reset` to `release_reset`.  That would just confuse folks.

**Review by bobsjune — COMMENTED** [review]
How do these new options appear in the espflash --help?

**Comment by nathan-whittington** [comment]
I'm not sure if it will cause any problems but I see three(?) resets when I exercise the `reset` command:
```
	ESP-ROM:esp32s3-20210327
	Build:Mar 27 2021
	rst:0x15 (USB_UART_CHIP_RESET),boot:0x8 (SPI_FAST_FLASH_BOOT)
	Saved PC:0x40378c52
	SPIWP:0xee
	mode:DIO, clock div:2
	load:0x3fce3810,len:0x1198
	load:0x403c9700,len:0xa50
	load:0x403cc700,len:0x2d88
	SHA-256 comparison failed:
	Calculated: b655b51fda251ec5ef85c33a9a8c1216e6db0cef46867cad3c9f5dbda0850921
	Expected: 7b22d32478d4df842a38e06a59debd2d38f6e76f524fdd78597a3546450207f0
	Attempting to boot anyway...
	entry 0x403c9894

	ESP-ROM:esp32s3-20210327
	Build:Mar 27 2021
	rst:0x15 (USB_UART_CHIP_RESET),boot:0x0 (DOWNLOAD(USB/UART0))
	Saved PC:0x40056f59
	waiting for download
	ESP-ROM:esp32s3-20210327
	Build:Mar 27 2021
	rst:0x15 (USB_UART_CHIP_RESET),boot:0x8 (SPI_FAST_FLASH_BOOT)
	Saved PC:0x40048df3
	SPIWP:0xee
	mode:DIO, clock div:2
	load:0x3fce3810,len:0x1198
	load:0x403c9700,len:0xa50
	load:0x403cc700,len:0x2d88
	SHA-256 comparison failed:
	Calculated: b655b51fda251ec5ef85c33a9a8c1216e6db0cef46867cad3c9f5dbda0850921
	Expected: 7b22d32478d4df842a38e06a59debd2d38f6e76f524fdd78597a3546450207f0
	Attempting to boot anyway...
	entry 0x403c9894
	I (229) esp_psram: Found 2MB PSRAM device
	I (229) esp_psram: Speed: 80MHz
	I (234) cpu_start: Pro cpu up.
	I (234) cpu_start: Starting app cpu, entry point is 0x40375fc4
	I (0) cpu_start: App cpu up.
```
I added the linefeed to help parse things.  

The three resets:
```
	rst:0x15 (USB_UART_CHIP_RESET),boot:0x8 (SPI_FAST_FLASH_BOOT)
	rst:0x15 (USB_UART_CHIP_RESET),boot:0x0 (DOWNLOAD(USB/UART0))
	rst:0x15 (USB_UART_CHIP_RESET),boot:0x8 (SPI_FAST_FLASH_BOOT)
```

For comparison I used esptool.py to reset the board (and I realized I don't know the command Anthony is using to do this).  

I used:
```
esptool.py --chip esp32s3 -p /dev/tty.usbmodem2121301 --before=default_reset --after=hard_reset read_flash 0 0x1000 ~/deleteme.bin
```
And this resulted in two resets:
```
ESP-ROM:esp32s3-20210327
Build:Mar 27 2021
rst:0x15 (USB_UART_CHIP_RESET),boot:0x0 (DOWNLOAD(USB/UART0))
Saved PC:0x40378c52
waiting for download

ESP-ROM:esp32s3-20210327
Build:Mar 27 2021
rst:0x15 (USB_UART_CHIP_RESET),boot:0x8 (SPI_FAST_FLASH_BOOT)
Saved PC:0x40378bb6
SPIWP:0xee
mode:DIO, clock div:2
load:0x3fce3810,len:0x1198
load:0x403c9700,len:0xa50
load:0x403cc700,len:0x2d88
SHA-256 comparison failed:
Calculated: b655b51fda251ec5ef85c33a9a8c1216e6db0cef46867cad3c9f5dbda0850921
Expected: 7b22d32478d4df842a38e06a59debd2d38f6e76f524fdd78597a3546450207f0
Attempting to boot anyway...
entry 0x403c9894
I (229) esp_psram: Found 2MB PSRAM device
I (229) esp_psram: Speed: 80MHz
I (234) cpu_start: Pro cpu up.
```

**Comment by nathan-whittington** [comment]
@bobsjune 

You can build from the local repo with `cargo build --release -p espflash -p cargo-espflash`

Install the new espflash from the local repo with:
```
cargo install --path ./espflash
cargo install --path ./cargo-espflash
```

**Comment by bobsjune** [comment]
I don't know why there would be three resets; I only expect two:  I'm guessing that in the first one, the ROM expects to load (based on the default strapping options) from SPI flash, but then "changes its mind", perhaps because it sees a connection on the DFU/UART; then the second is trying to download in DFU mode from the UART; that times out that and the ROM shifts back to reading from SPI flash (the DFU download would be the RAM stub).

**Comment by nathan-whittington** [comment]
Using hold-in-reset issues two resets: 
```
	rst:0x15 (USB_UART_CHIP_RESET),boot:0x8 (SPI_FAST_FLASH_BOOT)
	rst:0x15 (USB_UART_CHIP_RESET),boot:0x0 (DOWNLOAD(USB/UART0))
```
I think this is OK because it doesn't seem to get out of the ROM bootloader

**Review by nathan-whittington — APPROVED** [review]
I'd still like to see an option to prevent it from automatically resetting after a call to erase-region or write-bin but that can be a different PR.

**Comment by jnross** [comment]
> I'd still like to see an option to prevent it from automatically resetting after a call to erase-region or write-bin but that can be a different PR.

That's right.  That's simply a different request/desire.  This PR is about adding `reset` and `hold-in-reset` commands, not about adding options to other commands.

**Comment by jnross** [comment]
> I'm not sure if it will cause any problems but I see three(?) resets when I exercise the `reset` command:


I think I know why this is:  I believe the code here is loading the RAM stub that is used to flash or erase the device.  I don't believe this is necessary for the reset functionality.  

Which do you prefer:  Land this now and get the existing functionality which does perform a reset even if it's a bit extra?  Or wait until I or someone else can dig in further and refine the behavior?

**Comment by nathan-whittington** [comment]
I vote for land it now.  I don't think an extra reset causes us any grief.

**Comment by jnross** [comment]
I took another look at this.  We're already specifying `no_stub` in the reset command and it still performs these extra resets.  We'd have to untangle some code pretty deep in `espflash` to avoid this.  If it's important to you, let's make another ticket for it.

**Pull request merged**